wp_api/src/authenticator.rs

```diff
@@ -0,0 +1,310 @@
+    ) -> Option<HeaderMap> {
+        if self.should_authenticate(&request.url.0, Some(previous_response.status_code)) {
+            self.reset();
+            return self.authentication_headers().await;
+        }
+
+        None
+    }
+}
+
+#[derive(Debug)]
+pub(crate) struct NilAuthenticator {}
+
+#[async_trait::async_trait]
+impl Authenticator for NilAuthenticator {
+    fn should_authenticate(&self, _request_url: &str, _response_status_code: Option<u16>) -> bool {
+        false
+    }
+
+    fn host(&self) -> &str {
+        ""
+    }
+
+    async fn authentication_headers(&self) -> Option<HeaderMap> {
+        None
+    }
+
+    fn reset(&self) {
+        // Do nothing.
+    }
+}
+
+#[derive(Debug)]
+pub struct ApplicationPasswordAuthenticator {
+    host: String,
+    token: String,
+}
+
+impl ApplicationPasswordAuthenticator {
+    pub fn new(host: String, token: String) -> Self {
+        Self { host, token }
+    }
+
+    pub fn with_application_password(host: String, username: String, password: String) -> Self {
+        use base64::prelude::*;
+        let token = BASE64_STANDARD.encode(format!("{}:{}", username, password));
+        Self::new(host, token)
+    }
+}
+
+#[async_trait::async_trait]
+impl Authenticator for ApplicationPasswordAuthenticator {
+    fn host(&self) -> &str {
+        self.host.as_str()
+    }
+
+    async fn authentication_headers(&self) -> Option<HeaderMap> {
+        let mut headers = HeaderMap::new();
+        headers.append(
+            http::header::AUTHORIZATION,
+            format!("Basic {}", self.token)
+                .try_into()
+                .expect("token is always a valid header value"),
+        );
+
+        Some(headers)
+    }
+
+    fn reset(&self) {
+        // Do nothing.
+    }
+
+    fn should_authenticate(&self, request_url: &str, response_status_code: Option<u16>) -> bool {
+        let is_same_host = Url::parse(request_url)
+            .ok()
+            .map(|f| f.host_str() == Some(self.host()));
+        if is_same_host != Some(true) {
+            return false;
+        }
+
+        // Authenticated has already been sent. No need to make a second attempt.
+        response_status_code.is_none()
+    }
+}
+
+#[derive(Debug)]
+pub(crate) struct CookieAuthenticator {
+    api_base_url: ApiBaseUrl,
+    credentials: WpLoginCredentials,
+    request_executor: std::sync::Arc<dyn RequestExecutor>,
+    nonce: RwLock<Option<String>>,
+}
+
+impl CookieAuthenticator {
+    pub(crate) fn new(
+        api_base_url: ApiBaseUrl,
+        credentials: WpLoginCredentials,
+        request_executor: std::sync::Arc<dyn RequestExecutor>,
+    ) -> Self {
+        Self {
+            api_base_url,
+            credentials,
+            request_executor,
+            nonce: None.into(),
+        }
+    }
+
+    async fn get_rest_nonce(&self) -> Option<String> {
+        if let Some(cache) = self.nonce.read().expect("Failed to unlock nonce").clone() {
```

Comment:
Similar to the case where we are writing the `nonce`, there are valid cases where we can't read it. We can't just crash the app in such cases.